proposals/299-ip-failure-count.txt

```diff
@@ -92,26 +90,134 @@ Ticket: https://trac.torproject.org/projects/tor/ticket/27491
+      * IPv6 Failure Points: Count the number of IPv4-capable relays
+
+   If the consensus or bridge configuration changes during a session, we should
+   not update the failure point counters to generate a SFPV.
+
+   If we are starting a new session, we should use the existing failure points
+   to generate a SFPV unless the counts for IPv4 or IPv6 are zero.
+
+6. Forgetting Old Sessions
+
+   We should be able to forget old failures as clients could change networks.
+   For instance, a mobile phone could switch between WiFi and cellular. Keeping
+   an exact failure history would have privacy implications, so we should store
+   an approximate history.
+
+   One way we could forget old sessions is by halving all the failure point
+   (FP) values before adding when:
+
+      * One or more failure point values are a multiple of a random number
+        between 1 and 5
+
+      * One or more failure point values are greater than or equal to 100
+
+   The reason for halving the values at regular intervals is to forget old
+   sessions while keeping an approxmate history. We halve all FP values so
+   that one IP version doesn't dominante on the failure count if the other
+   is halved. This keeps an approximate scale of the failures on a client.
+
+   The reason for halving at a multiple of a random number instead of a fixed
+   interval is so we can halve regularly while not making it too predictable.
+   This prevents a situation where we would be halving too often to keep an
+   approximate failure history.
+
+   If we halve, we add the FP value for the failed IP version after halving all
+   FPs if done to account for the failure. If halving is not done, we will just
+   add the FP.
+
+   If the FP value for one IP version goes down to zero, we will re-calculate
+   the SFPV for that version using the methods described in Section 4.
+
+7. Separate Concurrent Connection Limits
+
+   Right now, there is a limit for three concurrent connections from a client.
+   at any given time. This limit includes both IPv4 and IPv6 connections.
+   This is to prevent denial of service attacks. I propose that a seperate
+   connection limit is used for IPv4 and IPv6. This means we can have three
+   concurrent IPv4 connections and three concurrent IPv6 connections at the
+   same time.
+
+   Having seperate connection limits allows us to deal with networks dropping
+   packets for a particular IP family while still preventing potential denial
+   of service attacks.
+
+8. Pathbias and Failure Probability
+
+   If ClientAutoIPv6ORPort is in use, and pathbias is triggered, we should
+   ignore "no route" warnings. The reason for this is because we would be
+   adding two failure points for the failed as described in Section 3 of this
+   proposal. Adding two failure points would make us more likely to prefer the
+   competing IP family over the failed one versus than adding a single failure
+   point on a normal failure.
+
+9. Counting Successful Connections
```

Comment:
I like this part of the design, too.

If a client is dual-stack, it may end up mostly preferring whichever protocol it tried first. That's probably ok.